webboot/webboot.go

```diff
@@ -23,9 +40,88 @@ var (
 	ipv6    = flag.Bool("ipv6", true, "use IPV6")
 )
 
+func wgetArgs(argURL string) (read io.ReadCloser, outFile string, werr error) {
+	//checks if argURL is a key in osSystems map
+	if _, ok := osSystems[argURL]; !ok {
+		//if argURL is not the map, use it as a URL link for wget
+		read, werr = wget(argURL)
+	} else {
+		// if argURL is in osSystems map, use its download link for wget
+		read, werr = wget(osSystems[argURL].DownloadLink)
+		outFile = argURL
+	}
+	return read, outFile, werr
+}
+
+func wget(arg string) (io.ReadCloser, error) {
+	resp, err := http.Get(arg)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("non-200 HTTP status: %d", resp.StatusCode)
+	}
+	return resp.Body, nil
+}
+
+func writer(read io.ReadCloser, fileName string) (string, error) {
+	w, err := os.Create(fileName)
+	if err != nil {
+		return "", err
+	}
+
+	defer w.Close()
+
+	_, err = io.Copy(w, read)
+	return fileName, err
+}
+
+func outputName(argURL string) (outFile string) {
+	//if output file is not initialized, argURl is parsed to generate a name for the output file
+	url, err := url.Parse(argURL)
+
+	if err != nil {
+		log.Fatalf("Error parsing the URL, got %v, want nil", err)
+	}
+	// takes the last part of the url Path as the name of output file
+	if url.Path != "" && url.Path[len(url.Path)-1] != '/' {
+		outFile = path.Base(url.Path)
+	} else {
+		outFile = "index.html"
+	}
+	return outFile
+}
+
+func usage() {
+	log.Printf("Usage: %s [ARGS] URL\n", os.Args[0])
+	flag.PrintDefaults()
+}
+
 func main() {
 	flag.Parse()
 
+	if flag.NArg() != 1 {
+		usage()
+		return
+	}
+
 	dhclient.Request(*ifName, *timeout, *retry, *verbose, *ipv4, *ipv6)
-	log.Println("weboot up")
+
+	argURL := flag.Arg(0)
+	read, outFile, werr := wgetArgs(argURL)
+	//checks the wget error
+	if werr != nil {
+		log.Fatalf("Error downloading the URL: %v", werr)
```

Comment:
error downloading what? You passed it back, why not tell me what you tried to get?
log.Fatalf("Error downloading %q: %v", u, err)

Reply:
and looking at your code, you return an URL, not a file, so see below.

Reply:
outFile is not a URL. it is a string that I use for local path to give to the folder that the downloaded contents gets stored to. See the comment above

Reply:
Passing the URL back to tell the user what URL i tried and failed with is a good idea. It will allow the user to see where the URL failure went wrong and with what it went wrong with.  I like that idea. I pass the URL back from the function which is now called parseArgs(). Although, I print the failure message in regards to the url and customize the error messages in each error case within the parseArgs() function and just log.Fatal(err) in the main function to let the customized errors I have created in the function speak for themselves.

Reply:
I think there is a slight confusion with the variable "outFile". outFile is not a URL within my code. It is simply a name of the local output file name. For example, "webboot TinyCore", then outFile = "TinyCore". If a user says "webboot http://mirrors.edge.kernel.org/archlinux/iso/2019.08.01/archlinux-2019.08.01-x86_64.iso" then the outFile uses the base of the link to name the local output file

Reply:
I think there is a slight confusion with the variable "outFile". outFile is not a URL within my code. It is simply a name of the local output file name. For example, "webboot TinyCore", then outFile = "TinyCore". If a user says "webboot http://mirrors.edge.kernel.org/archlinux/iso/2019.08.01/archlinux-2019.08.01-x86_64.iso" then the outFile uses the base of the link to name the local output file, so in this case outFile = archlinux-2019.08.01-x86_64.iso

Reply:
I think there is a slight confusion with the variable "outFile". outFile is not a URL within my code. It is the local filename given to the file that stores all of the downloaded content. For example, "webboot TinyCore", then outFile = "TinyCore". If a user says "webboot http://mirrors.edge.kernel.org/archlinux/iso/2019.08.01/archlinux-2019.08.01-x86_64.iso" then the outFile uses the base of the link to name the local output file, so in this case outFile = archlinux-2019.08.01-x86_64.iso